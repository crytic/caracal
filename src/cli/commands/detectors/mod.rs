use super::Cmd;
use caracal::detectors::get_detectors;
use clap::Args;

#[derive(Args, Debug)]
pub struct DetectorsArgs {}

impl Cmd for DetectorsArgs {
    fn run(&self) -> anyhow::Result<()> {
        get_detectors().iter().for_each(|d| println!("{}", d));
        Ok(())
    }
}
