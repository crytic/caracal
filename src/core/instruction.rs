use cairo_lang_sierra::{
    ids::VarId,
    program::{GenStatement, Statement as SierraStatement},
};

#[derive(Debug, Clone)]
pub struct Instruction {
    pc: usize,
    statement: SierraStatement,
}

impl Instruction {
    pub fn new(pc: usize, statement: SierraStatement) -> Self {
        Instruction { pc, statement }
    }

    pub fn get_pc(&self) -> usize {
        self.pc
    }

    pub fn get_statement(&self) -> &SierraStatement {
        &self.statement
    }

    pub fn variables_read(&self) -> &Vec<VarId> {
        match &self.statement {
            GenStatement::Invocation(inv) => &inv.args,
            GenStatement::Return(vars) => vars,
        }
    }

    pub fn variables_written(&self) -> Vec<VarId> {
        match &self.statement {
            GenStatement::Invocation(inv) => {
                let mut vars_written = Vec::new();
                for branch in inv.branches.iter() {
                    vars_written.extend(branch.results.clone());
                }
                vars_written
            }
            GenStatement::Return(_) => {
                vec![]
            }
        }
    }
}

impl std::fmt::Display for Instruction {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} {}", self.pc, self.statement)
    }
}
