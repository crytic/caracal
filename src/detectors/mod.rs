use self::detector::Detector;

pub mod array_use_after_pop_front;
pub mod controlled_library_call;
pub mod dead_code;
pub mod detector;
pub mod read_only_reentrancy;
pub mod reentrancy;
pub mod reentrancy_benign;
pub mod reentrancy_events;
pub mod unchecked_l1_handler_from;
pub mod unused_arguments;
pub mod unused_events;
pub mod unused_return;

pub fn get_detectors() -> Vec<Box<dyn Detector>> {
    vec![
        Box::<array_use_after_pop_front::ArrayUseAfterPopFront>::default(),
        Box::<controlled_library_call::ControlledLibraryCall>::default(),
        Box::<unused_events::UnusedEvents>::default(),
        Box::<dead_code::DeadCode>::default(),
        Box::<unused_arguments::UnusedArguments>::default(),
        Box::<unused_return::UnusedReturn>::default(),
        Box::<reentrancy_benign::ReentrancyBenign>::default(),
        Box::<reentrancy::Reentrancy>::default(),
        Box::<reentrancy_events::ReentrancyEvents>::default(),
        Box::<read_only_reentrancy::ReadOnlyReentrancy>::default(),
        Box::<unchecked_l1_handler_from::UncheckedL1HandlerFrom>::default(),
    ]
}
