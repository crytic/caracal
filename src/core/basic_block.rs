use super::instruction::Instruction;
use crate::core::function::{Function, Type};
use std::hash::{Hash, Hasher};

use cairo_lang_sierra::extensions::core::{CoreConcreteLibfunc, CoreLibfunc, CoreType};
use cairo_lang_sierra::program::GenStatement;
use cairo_lang_sierra::program_registry::ProgramRegistry;

#[derive(Debug, Clone, Default)]
pub struct BasicBlock {
    function: String,
    id: usize,
    instructions: Vec<Instruction>,
    incoming_basic_blocks: Vec<usize>,
    outgoing_basic_blocks: Vec<usize>,
    private_call: Option<Instruction>,
    external_call: Option<Instruction>,
    library_call: Option<Instruction>,
    storage_variable_read: Option<Instruction>,
    storage_variable_written: Option<Instruction>,
    event_emitted: Option<Instruction>,
}

impl PartialEq for BasicBlock {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}

impl Eq for BasicBlock {}

impl Hash for BasicBlock {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.function.hash(state);
        self.id.hash(state);
    }
}

impl BasicBlock {
    pub fn new(
        function: String,
        id: usize,
        instructions: Vec<Instruction>,
        functions: &[Function],
        registry: &ProgramRegistry<CoreType, CoreLibfunc>,
    ) -> Self {
        let mut bb = BasicBlock {
            function,
            id,
            instructions,
            ..Default::default()
        };

        bb.analyze(functions, registry);

        bb
    }

    pub fn len(&self) -> usize {
        self.instructions.len()
    }

    pub fn is_empty(&self) -> bool {
        self.instructions.is_empty()
    }

    pub fn get_id(&self) -> usize {
        self.id
    }

    pub fn get_function(&self) -> &str {
        &self.function
    }

    pub fn get_instructions(&self) -> &Vec<Instruction> {
        &self.instructions
    }

    pub fn get_incoming_basic_blocks(&self) -> &Vec<usize> {
        &self.incoming_basic_blocks
    }

    pub fn get_outgoing_basic_blocks(&self) -> &Vec<usize> {
        &self.outgoing_basic_blocks
    }

    /// Return the function call in this basic block
    pub fn get_function_call(&self) -> Option<&Instruction> {
        // Assume there can be only one function call in each basic block
        // when a function call is executed sierra always check for panics so it should be true

        if let Some(instruction) = &self.event_emitted {
            return Some(instruction);
        }
        if let Some(instruction) = &self.external_call {
            return Some(instruction);
        }
        if let Some(instruction) = &self.library_call {
            return Some(instruction);
        }
        if let Some(instruction) = &self.private_call {
            return Some(instruction);
        }
        if let Some(instruction) = &self.storage_variable_read {
            return Some(instruction);
        }
        if let Some(instruction) = &self.storage_variable_written {
            return Some(instruction);
        }

        None
    }

    pub fn add_incoming_bb(&mut self, basic_block: usize) {
        self.incoming_basic_blocks.push(basic_block);
    }

    pub fn add_outgoing_bb(&mut self, basic_block: usize) {
        self.outgoing_basic_blocks.push(basic_block);
    }

    pub fn remove_incoming_bb(&mut self, basic_block: usize) {
        let pos = self
            .incoming_basic_blocks
            .iter()
            .position(|id| *id == basic_block)
            .expect("Trying to remove an incoming edge that doesn't exist");
        self.incoming_basic_blocks.remove(pos);
    }

    pub fn remove_outgoing_bb(&mut self, basic_block: usize) {
        let pos = self
            .outgoing_basic_blocks
            .iter()
            .position(|id| *id == basic_block)
            .expect("Trying to remove an outgoing edge that doesn't exist");
        self.outgoing_basic_blocks.remove(pos);
    }

    pub fn last_instruction(&self) -> Option<&Instruction> {
        self.instructions.last()
    }

    pub fn first_instruction(&self) -> Option<&Instruction> {
        self.instructions.first()
    }

    pub fn get_private_call(&self) -> &Option<Instruction> {
        &self.private_call
    }

    pub fn get_external_call(&self) -> &Option<Instruction> {
        &self.external_call
    }

    pub fn get_library_call(&self) -> &Option<Instruction> {
        &self.library_call
    }

    pub fn get_storage_variable_read(&self) -> &Option<Instruction> {
        &self.storage_variable_read
    }

    pub fn get_storage_variable_written(&self) -> &Option<Instruction> {
        &self.storage_variable_written
    }

    pub fn get_event_emitted(&self) -> &Option<Instruction> {
        &self.event_emitted
    }

    pub fn analyze(
        &mut self,
        functions: &[Function],
        registry: &ProgramRegistry<CoreType, CoreLibfunc>,
    ) {
        for instruction in self.instructions.iter() {
            if let GenStatement::Invocation(invoc) = instruction.get_statement() {
                let lib_func = registry
                    .get_libfunc(&invoc.libfunc_id)
                    .expect("Library function not found in the registry");

                if let CoreConcreteLibfunc::FunctionCall(f_called) = lib_func {
                    // We search for the function called in our list of functions to know its type
                    for function in functions {
                        let function_name = function.name();
                        if function_name.as_str()
                            == f_called.function.id.debug_name.as_ref().unwrap()
                        {
                            match function.ty() {
                                Type::Storage => {
                                    if function_name.ends_with("::read") {
                                        self.storage_variable_read = Some(instruction.clone());
                                    } else if function_name.ends_with("::write") {
                                        self.storage_variable_written = Some(instruction.clone());
                                    }
                                }
                                Type::Event => self.event_emitted = Some(instruction.clone()),
                                Type::Private => self.private_call = Some(instruction.clone()),
                                Type::AbiCallContract => {
                                    self.external_call = Some(instruction.clone())
                                }
                                Type::AbiLibraryCall => {
                                    self.library_call = Some(instruction.clone())
                                }
                                _ => (),
                            }
                            break;
                        }
                    }
                }
            }
        }
    }
}
