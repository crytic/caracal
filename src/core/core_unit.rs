use crate::compilation::compile;
use crate::core::compilation_unit::CompilationUnit;
use anyhow::Result;
use cairo_lang_sierra::extensions::core::{CoreLibfunc, CoreType};
use cairo_lang_sierra::program_registry::ProgramRegistry;
use rayon::prelude::*;
use std::path::PathBuf;

pub struct CoreOpts {
    pub target: PathBuf,
    pub corelib: Option<PathBuf>,
    pub contract_path: Option<Vec<String>>,
    pub safe_external_calls: Option<Vec<String>>,
}

pub struct CoreUnit {
    compilation_units: Vec<CompilationUnit>,
    safe_external_calls: Option<Vec<String>>,
}

impl CoreUnit {
    pub fn new(opts: CoreOpts) -> Result<Self> {
        let safe_external_calls = opts.safe_external_calls.clone();
        let program_compiled = compile(opts)?;
        let compilation_units = program_compiled
            .par_iter()
            .map(|p| {
                let mut compilation_unit = CompilationUnit::new(
                    p.sierra.clone(),
                    p.abi.clone(),
                    ProgramRegistry::<CoreType, CoreLibfunc>::new(&p.sierra).unwrap(),
                );
                compilation_unit.analyze();
                compilation_unit
            })
            .collect();
        Ok(CoreUnit {
            compilation_units,
            safe_external_calls,
        })
    }

    pub fn get_compilation_units(&self) -> &Vec<CompilationUnit> {
        &self.compilation_units
    }

    pub fn get_safe_external_calls(&self) -> &Option<Vec<String>> {
        &self.safe_external_calls
    }
}
