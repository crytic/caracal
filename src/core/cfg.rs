use std::collections::{HashMap, HashSet};

use super::basic_block::BasicBlock;
use super::function::Function;
use super::instruction::Instruction;
use cairo_lang_sierra::extensions::core::{CoreLibfunc, CoreType};
use cairo_lang_sierra::program::{BranchTarget, Statement as SierraStatement};
use cairo_lang_sierra::program_registry::ProgramRegistry;

pub trait Cfg {
    fn get_basic_blocks(&self) -> &[BasicBlock];
    fn get_basic_block(&self, id: usize) -> Option<&BasicBlock>;
}

#[derive(Debug, Clone, Default)]
pub struct CfgRegular {
    basic_blocks: Vec<BasicBlock>,
}

impl CfgRegular {
    pub fn new() -> Self {
        CfgRegular {
            basic_blocks: Vec::new(),
        }
    }

    pub fn analyze(
        &mut self,
        statements: &[SierraStatement],
        base_pc: usize,
        functions: &[Function],
        registry: &ProgramRegistry<CoreType, CoreLibfunc>,
        function_name: String,
    ) {
        self.compute_basic_blocks(statements, base_pc, functions, registry, function_name);
        self.compute_cfg();
    }

    fn compute_basic_blocks(
        &mut self,
        statements: &[SierraStatement],
        base_pc: usize,
        functions: &[Function],
        registry: &ProgramRegistry<CoreType, CoreLibfunc>,
        function_name: String,
    ) {
        // Track basic block ids
        let mut basic_block_counter = 0;
        let mut basic_blocks = Vec::new();

        // The instructions of the current block
        let mut instructions_current_block = Vec::new();

        // Tracks PCs of jump's target
        let mut target_pcs: HashSet<usize> = HashSet::new();

        for (i, statement) in statements.iter().enumerate() {
            let current_pc = base_pc + i;
            match statement {
                // Statement with a single branch. It's a statemnt that doesn't change the flow (Fallthrough) or an unconditional jump
                SierraStatement::Invocation(function) if function.branches.len() == 1 => {
                    match function.branches[0].target {
                        BranchTarget::Fallthrough => {
                            self.handle_statement_basic_block(
                                statement.clone(),
                                &target_pcs,
                                current_pc,
                                &mut basic_blocks,
                                &mut instructions_current_block,
                                &mut basic_block_counter,
                                functions,
                                registry,
                                function_name.clone(),
                            );
                        }
                        BranchTarget::Statement(pc) => {
                            // Unconditional jump
                            target_pcs.insert(pc.0);
                            self.handle_statement_basic_block(
                                statement.clone(),
                                &target_pcs,
                                current_pc,
                                &mut basic_blocks,
                                &mut instructions_current_block,
                                &mut basic_block_counter,
                                functions,
                                registry,
                                function_name.clone(),
                            );
                        }
                    }
                }
                // Statement with multiple branches
                SierraStatement::Invocation(function) => {
                    for branch in function.branches.iter() {
                        match branch.target {
                            BranchTarget::Statement(pc) => {
                                target_pcs.insert(pc.0);
                            }
                            BranchTarget::Fallthrough => {
                                target_pcs.insert(current_pc + 1);
                            }
                        }
                    }

                    self.handle_statement_basic_block(
                        statement.clone(),
                        &target_pcs,
                        current_pc,
                        &mut basic_blocks,
                        &mut instructions_current_block,
                        &mut basic_block_counter,
                        functions,
                        registry,
                        function_name.clone(),
                    );
                }
                SierraStatement::Return(_) => {
                    // Always terminate the current block
                    // Add the instruction in the current block
                    instructions_current_block
                        .push(Instruction::new(current_pc, statement.clone()));
                    // Create a new basic block
                    basic_blocks.push(BasicBlock::new(
                        function_name.clone(),
                        basic_block_counter,
                        instructions_current_block.clone(),
                        functions,
                        registry,
                    ));
                    // Clear the current instructions for the next basic block
                    instructions_current_block.clear();
                    basic_block_counter += 1;
                }
            }
        }

        self.basic_blocks = basic_blocks;
    }

    #[allow(clippy::too_many_arguments)]
    fn handle_statement_basic_block(
        &self,
        statement: SierraStatement,
        target_pcs: &HashSet<usize>,
        current_pc: usize,
        basic_blocks: &mut Vec<BasicBlock>,
        instructions_current_block: &mut Vec<Instruction>,
        basic_block_counter: &mut usize,
        functions: &[Function],
        registry: &ProgramRegistry<CoreType, CoreLibfunc>,
        function_name: String,
    ) {
        // Add the instruction in the current block
        instructions_current_block.push(Instruction::new(current_pc, statement));

        // If the current statement is a target of a jump complete the current basic block
        if target_pcs.get(&(current_pc + 1)).is_some() {
            // Create a new basic block
            basic_blocks.push(BasicBlock::new(
                function_name,
                *basic_block_counter,
                instructions_current_block.clone(),
                functions,
                registry,
            ));
            // Clear the current instructions for the next basic block
            instructions_current_block.clear();
            *basic_block_counter += 1;
        }
    }

    fn compute_cfg(&mut self) {
        // Track the incoming edges - basic block id -> incoming basic blocks id
        let mut incoming: HashMap<usize, Vec<usize>> = HashMap::new();
        // Track the outgoing edges - basic block id -> outgoing basic blocks id
        let mut outgoing: HashMap<usize, Vec<usize>> = HashMap::new();

        for source_bb in self.basic_blocks.iter() {
            let last_instruction = source_bb
                .last_instruction()
                .expect("Basic Block with 0 instruction");
            if let SierraStatement::Invocation(function) = &last_instruction.get_statement() {
                for branch in function.branches.iter() {
                    match branch.target {
                        BranchTarget::Statement(destination) => {
                            self.handle_statement_cfg(
                                destination.0,
                                &mut incoming,
                                &mut outgoing,
                                source_bb.get_id(),
                            );
                        }
                        // Fallthrough, the target_pc is the subsequent PC
                        BranchTarget::Fallthrough => {
                            self.handle_statement_cfg(
                                last_instruction.get_pc() + 1,
                                &mut incoming,
                                &mut outgoing,
                                source_bb.get_id(),
                            );
                        }
                    }
                }
            }
        }

        for bb in self.basic_blocks.iter_mut() {
            // Add the outgoing basic block for the bb if any
            if let Some(destinations) = outgoing.get(&bb.get_id()) {
                destinations.iter().for_each(|d| bb.add_outgoing_bb(*d));
            }
            // Add the incoming basic block for the bb if any
            if let Some(sources) = incoming.get(&bb.get_id()) {
                sources.iter().for_each(|s| bb.add_incoming_bb(*s));
            }
        }
    }

    // Statement that can be a target of a jump
    fn handle_statement_cfg(
        &self,
        target_pc: usize,
        incoming: &mut HashMap<usize, Vec<usize>>,
        outgoing: &mut HashMap<usize, Vec<usize>>,
        source_bb_id: usize,
    ) {
        for bb in self.basic_blocks.iter() {
            if bb
                .first_instruction()
                .expect("Basic Block with 0 instruction")
                .get_pc()
                == target_pc
            {
                let destination_bb_id = bb.get_id();

                let sources = incoming.entry(destination_bb_id).or_default();
                sources.push(source_bb_id);
                let destinations = outgoing.entry(source_bb_id).or_default();
                destinations.push(destination_bb_id);
                break;
            }
        }
    }
}

impl Cfg for CfgRegular {
    fn get_basic_blocks(&self) -> &[BasicBlock] {
        &self.basic_blocks
    }

    fn get_basic_block(&self, id: usize) -> Option<&BasicBlock> {
        self.basic_blocks.get(id)
    }
}

#[derive(Debug, Clone, Default)]
pub struct CfgOptimized {
    basic_blocks: Vec<BasicBlock>,
}

impl CfgOptimized {
    pub fn new() -> Self {
        CfgOptimized {
            basic_blocks: Vec::new(),
        }
    }

    pub fn analyze(&mut self, basic_blocks: Vec<BasicBlock>) {
        self.basic_blocks = basic_blocks;
        self.remove_redundant_bbs();
        self.rename_bbs();
    }

    fn get_mut_basic_block(&mut self, id: usize) -> Option<&mut BasicBlock> {
        self.basic_blocks.iter_mut().find(|bb| bb.get_id() == id)
    }

    // Should be used during optimization passes because the basic blocks may not be in the correct sequence
    // which means when we remove basic_blocks from [1,2,3,4] becomes [1,3,4]
    // so we can't just get the basic block at id
    fn get_basic_block_not_renamed(&self, id: usize) -> Option<&BasicBlock> {
        self.basic_blocks.iter().find(|bb| bb.get_id() == id)
    }

    /// Merge a basic block with 1 outgoing basic block to the following basic block if it has a single incoming basic block
    fn remove_redundant_bbs(&mut self) {
        // source basic block id | destination basic block id | destination basic block's outgoing edges to add in the source basic block | instructions to add in the source basic block
        let mut bb_to_merge: Vec<(usize, usize, Vec<usize>, Vec<Instruction>)> = Vec::new();

        for bb in self.basic_blocks.iter() {
            if bb.get_outgoing_basic_blocks().len() == 1 {
                let outgoing_bb = self
                    .get_basic_block_not_renamed(bb.get_outgoing_basic_blocks()[0])
                    .expect("Outgoing basic block not found");
                if outgoing_bb.get_incoming_basic_blocks().len() == 1 {
                    bb_to_merge.push((
                        bb.get_id(),
                        outgoing_bb.get_id(),
                        outgoing_bb.get_outgoing_basic_blocks().clone(),
                        outgoing_bb.get_instructions().clone(),
                    ));
                }
            }
        }

        for (bb_source_id, bb_destination_id, new_outgoing_bbs, new_instructions) in
            bb_to_merge.iter()
        {
            let bb_to_mutate = self
                .get_mut_basic_block(*bb_source_id)
                .expect("Source basic block not found");
            bb_to_mutate.remove_outgoing_bb(*bb_destination_id);
            for bb in new_outgoing_bbs {
                bb_to_mutate.add_outgoing_bb(*bb);
            }
            // Remove the last instruction
            // We assume it's a jump, maybe we could check it
            bb_to_mutate.get_mut_instructions().pop();
            // Add the destination basic block instructions
            bb_to_mutate
                .get_mut_instructions()
                .extend(new_instructions.clone());
        }

        // Update incoming edges for the outgoing basic block of the removed basic block
        for (bb_source_id, bb_destination_id, new_outgoing_bbs, _) in bb_to_merge.iter() {
            let bb_index_to_remove = self
                .basic_blocks
                .iter()
                .position(|bb| bb.get_id() == *bb_destination_id)
                .expect("Didn't find basic block to remove");
            self.basic_blocks.remove(bb_index_to_remove);
            for bb in new_outgoing_bbs {
                // Delete the incoming edge from the removed basic block and add bb_source_id
                let bb_to_mutate = self
                    .get_mut_basic_block(*bb)
                    .expect("New outgoing basic block not found");
                bb_to_mutate.remove_incoming_bb(*bb_destination_id);
                bb_to_mutate.add_incoming_bb(*bb_source_id);
            }
        }
    }

    fn rename_bbs(&mut self) {
        // Track how many basic block were removed and then subtract that value from the next basic block id
        let mut bb_counter_to_subtract = 0;
        // Track bb id => new id
        let mut bb_to_new_id = HashMap::new();

        for (i, bb) in self.basic_blocks.iter().enumerate() {
            let bb_id = bb.get_id();
            // Check if the bb_id is in the correct sequence
            if i != bb_id {
                // A basic block was removed between this block and the previous block
                // we need to increment the bb_counter_to_subtract
                // [0,1,2,3]
                // [0,2,3]
                //    ^
                // 2 - 1 > 0 -> bb_counter_to_subtract += 1
                // new id = 2 - 1 = 1
                // [0,1,3]
                //      ^
                // 3 - 2 > 1 -> No
                // new id = 3 - 1 = 2
                // [0,1,2]
                if bb_id - i > bb_counter_to_subtract {
                    bb_counter_to_subtract += 1;
                }
                let bb_new_id = bb_id - bb_counter_to_subtract;
                bb_to_new_id.insert(bb_id, bb_new_id);
            }
        }

        // Update the id, incoming and outgoing basic blocks
        for bb in self.basic_blocks.iter_mut() {
            if let Some(new_id) = bb_to_new_id.get(&bb.get_id()) {
                let bb_id = bb.get_mut_id();
                *bb_id = *new_id;
            }

            for incoming_bb in bb.get_incoming_basic_blocks().clone() {
                if let Some(new_id) = bb_to_new_id.get(&incoming_bb) {
                    bb.remove_incoming_bb(incoming_bb);
                    bb.add_incoming_bb(*new_id);
                }
            }
            for outgoing_bb in bb.get_outgoing_basic_blocks().clone() {
                if let Some(new_id) = bb_to_new_id.get(&outgoing_bb) {
                    bb.remove_outgoing_bb(outgoing_bb);
                    bb.add_outgoing_bb(*new_id);
                }
            }
        }
    }
}

impl Cfg for CfgOptimized {
    fn get_basic_blocks(&self) -> &[BasicBlock] {
        &self.basic_blocks
    }

    fn get_basic_block(&self, id: usize) -> Option<&BasicBlock> {
        self.basic_blocks.get(id)
    }
}
