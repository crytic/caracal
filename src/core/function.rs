use std::io::Write;

use super::cfg::{Cfg, CfgOptimized, CfgRegular};
use crate::utils::BUILTINS;
use cairo_lang_sierra::extensions::core::{CoreConcreteLibfunc, CoreLibfunc, CoreType};
use cairo_lang_sierra::ids::ConcreteTypeId;
use cairo_lang_sierra::program::{
    Function as SierraFunction, GenStatement, Param, Statement as SierraStatement,
};
use cairo_lang_sierra::program_registry::ProgramRegistry;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Type {
    /// External function defined by the user
    External,
    /// View function defined by the user
    View,
    /// Private function defined by the user
    Private,
    /// Constructor function defined by the user
    Constructor,
    /// Event function
    Event,
    /// Function made by the compiler for storage variables
    /// typically address, read, write
    Storage,
    /// Wrapper around an external function made by the compiler
    Wrapper,
    /// Function of the core library
    Core,
    /// Function of a trait with the ABI attribute that does a call contract
    AbiCallContract,
    /// Function of a trait with the ABI attribute that does a library call
    AbiLibraryCall,
    /// L1 handler function
    L1Handler,
}

#[derive(Clone)]
pub struct Function {
    /// Underlying Function data
    data: SierraFunction,
    /// Type of function
    ty: Option<Type>,
    /// The sequence of statements
    statements: Vec<SierraStatement>,
    /// A regular CFG from the statements
    cfg_regular: CfgRegular,
    /// An optimized CFG from the statements
    cfg_optimized: CfgOptimized,
    /// Storage variables read (NOTE it doesn't have vars read using the syscall directly)
    storage_vars_read: Vec<SierraStatement>,
    /// Storage variables written (NOTE it doesn't have vars written using the syscall directly)
    storage_vars_written: Vec<SierraStatement>,
    /// Core functions called
    core_functions_calls: Vec<SierraStatement>,
    /// Private functions called
    private_functions_calls: Vec<SierraStatement>,
    /// Events emitted (NOTE it doesn't have events emitted using the syscall directly)
    events_emitted: Vec<SierraStatement>,
    /// External functions called through an ABI trait (NOTE it doesn't have external functions called using the syscall directly)
    external_functions_calls: Vec<SierraStatement>,
    /// Library functions called through an ABI trait (NOTE it doesn't have library functions called using the syscall directly)
    library_functions_calls: Vec<SierraStatement>,
}

impl Function {
    pub fn new(data: SierraFunction, statements: Vec<SierraStatement>) -> Self {
        Function {
            data,
            ty: None,
            statements,
            cfg_regular: CfgRegular::new(),
            cfg_optimized: CfgOptimized::new(),
            storage_vars_read: Vec::new(),
            storage_vars_written: Vec::new(),
            core_functions_calls: Vec::new(),
            private_functions_calls: Vec::new(),
            events_emitted: Vec::new(),
            external_functions_calls: Vec::new(),
            library_functions_calls: Vec::new(),
        }
    }

    pub fn name(&self) -> String {
        self.data.id.to_string()
    }

    pub fn ty(&self) -> &Type {
        // At this point is always initialized
        self.ty.as_ref().unwrap()
    }

    pub fn storage_vars_read(&self) -> impl Iterator<Item = &SierraStatement> {
        self.storage_vars_read.iter()
    }

    pub fn storage_vars_written(&self) -> impl Iterator<Item = &SierraStatement> {
        self.storage_vars_written.iter()
    }

    pub fn core_functions_calls(&self) -> impl Iterator<Item = &SierraStatement> {
        self.core_functions_calls.iter()
    }

    pub fn private_functions_calls(&self) -> impl Iterator<Item = &SierraStatement> {
        self.private_functions_calls.iter()
    }

    pub fn events_emitted(&self) -> impl Iterator<Item = &SierraStatement> {
        self.events_emitted.iter()
    }

    pub fn external_functions_calls(&self) -> impl Iterator<Item = &SierraStatement> {
        self.external_functions_calls.iter()
    }

    pub fn library_functions_calls(&self) -> impl Iterator<Item = &SierraStatement> {
        self.library_functions_calls.iter()
    }

    /// Function return variables without the builtins
    pub fn returns(&self) -> impl Iterator<Item = &ConcreteTypeId> {
        self.data
            .signature
            .ret_types
            .iter()
            .filter(|r| !BUILTINS.contains(&r.debug_name.clone().unwrap().as_str()))
    }

    /// Function return variables
    pub fn returns_all(&self) -> impl Iterator<Item = &ConcreteTypeId> {
        self.data.signature.ret_types.iter()
    }

    /// Function parameters without the builtins
    pub fn params(&self) -> impl Iterator<Item = &Param> {
        self.data
            .params
            .iter()
            .filter(|p| !BUILTINS.contains(&p.ty.debug_name.clone().unwrap().as_str()))
    }

    /// Function parameters
    pub fn params_all(&self) -> impl Iterator<Item = &Param> {
        self.data.params.iter()
    }

    pub fn get_statements(&self) -> &Vec<SierraStatement> {
        &self.statements
    }

    pub fn get_statements_at(&self, at: usize) -> &[SierraStatement] {
        &self.statements[at..]
    }

    pub fn get_cfg(&self) -> &CfgRegular {
        &self.cfg_regular
    }

    pub fn get_cfg_optimized(&self) -> &CfgOptimized {
        &self.cfg_optimized
    }

    pub fn analyze(
        &mut self,
        functions: &[Function],
        registry: &ProgramRegistry<CoreType, CoreLibfunc>,
    ) {
        self.cfg_regular
            .analyze(&self.statements, self.data.entry_point.0);
        self.cfg_optimized
            .analyze(self.cfg_regular.get_basic_blocks().to_vec());
        self.set_meta_informations(functions, registry);
    }

    /// Set the meta informations such as storage variables read, storage variables written, core function called
    /// private function called, events emitted
    fn set_meta_informations(
        &mut self,
        functions: &[Function],
        registry: &ProgramRegistry<CoreType, CoreLibfunc>,
    ) {
        for s in self.statements.iter() {
            if let GenStatement::Invocation(invoc) = s {
                let lib_func = registry
                    .get_libfunc(&invoc.libfunc_id)
                    .expect("Library function not found in the registry");
                if let CoreConcreteLibfunc::FunctionCall(f_called) = lib_func {
                    // We search for the function called in our list of functions to know its type
                    for function in functions.iter() {
                        let function_name = function.name();
                        if function_name.as_str()
                            == f_called.function.id.debug_name.as_ref().unwrap()
                        {
                            match function.ty() {
                                Type::Storage => {
                                    if function_name.ends_with("read") {
                                        self.storage_vars_read.push(s.clone());
                                    } else if function_name.ends_with("write") {
                                        self.storage_vars_written.push(s.clone());
                                    }
                                }
                                Type::Event => self.events_emitted.push(s.clone()),
                                Type::Core => self.core_functions_calls.push(s.clone()),
                                Type::Private => self.private_functions_calls.push(s.clone()),
                                Type::AbiCallContract => {
                                    self.external_functions_calls.push(s.clone())
                                }
                                Type::AbiLibraryCall => {
                                    self.library_functions_calls.push(s.clone())
                                }
                                _ => (),
                            }
                            break;
                        }
                    }
                }
            }
        }
    }

    pub(super) fn set_ty(&mut self, ty: Type) {
        self.ty = Some(ty);
    }

    /// Write to a file the function's CFG and return the filename
    pub fn cfg_to_dot(&self, cfg: &dyn Cfg) -> String {
        // name for now good enough
        let file_name = format!(
            "{}.dot",
            self.name()
                .split('<')
                .take(1)
                .next()
                .expect("Error when creating the filename")
        )
        .replace("::", "_");

        let mut f = std::fs::OpenOptions::new()
            .write(true)
            .truncate(true)
            .create(true)
            .open(&file_name)
            .expect("Error when creating file");

        f.write_all(b"digraph{\n").unwrap();

        for bb in cfg.get_basic_blocks() {
            let mut ins = String::new();
            bb.get_instructions()
                .iter()
                .for_each(|i| ins.push_str(&format!("{i}\n")));
            f.write_all(
                format!("{}[label=\"BB {}\n{}\"]\n", bb.get_id(), bb.get_id(), ins).as_bytes(),
            )
            .unwrap();

            for destination in bb.get_outgoing_basic_blocks().iter() {
                f.write_all(format!("{} -> {}\n", bb.get_id(), destination).as_bytes())
                    .unwrap();
            }
        }

        f.write_all(b"}\n").unwrap();

        file_name
    }
}
