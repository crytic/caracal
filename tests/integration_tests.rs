use starknet_static_analysis::core::core_unit::{CoreOpts, CoreUnit};
use starknet_static_analysis::detectors::{detector::Result, get_detectors};
use std::env;
use std::path::PathBuf;

#[test]
fn test_detectors() {
    insta::glob!("detectors/", "*.cairo", |path| {
        let opts = CoreOpts {
            target: path.to_path_buf(),
            corelib: Some(PathBuf::from(
                env::var("CARGO_MANIFEST_DIR").unwrap() + "/corelib/src",
            )),
        };
        let core = CoreUnit::new(opts).unwrap();
        insta::assert_debug_snapshot!(get_detectors()
            .iter()
            .flat_map(|d| d.run(&core))
            .collect::<Vec<Result>>());
    });
}
