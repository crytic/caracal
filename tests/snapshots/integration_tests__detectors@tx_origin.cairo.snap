---
source: tests/integration_tests.rs
expression: results
input_file: tests/detectors/tx_origin.cairo
---
[
    Result {
        impact: High,
        name: "dangerous-use-of-transaction-origin",
        confidence: Medium,
        message: "The transaction origin contract address is used in an authentication check in the function tx_origin::tx_origin::TxOrigin::bad",
    },
    Result {
        impact: High,
        name: "dangerous-use-of-transaction-origin",
        confidence: Medium,
        message: "The transaction origin contract address is used in an authentication check in the function tx_origin::tx_origin::TxOrigin::bad_indirect",
    },
]
